File: source/client/src/api/pixcavationApi.js
const BASE_URL = "";

async function apiFetch(path, {method = "GET", body} = {}) {
    const res = await fetch(`${BASE_URL}${path}`, {
        method,
        headers: {"Content-Type": "application/json"},
        body: body ? JSON.stringify(body) : undefined,
    });

    const contentType = res.headers.get("content-type") || "";
    const data = contentType.includes("application/json") ? await res.json() : await res.text();

    if (!res.ok) throw new Error(data?.error || `Request failed (${res.status})`);
    if (data && typeof data === "object" && data.ok === false) throw new Error(data.error || "Request failed");

    return data;
}

export const pixcavationApi = {
    exists() {
        return apiFetch("/api/game/exists");
    },
    newGame() {
        return apiFetch("/api/game/new", {method: "POST"});
    },
    getGame() {
        return apiFetch("/api/game");
    },
    // revealPixel(x, y) {
    //     return apiFetch("/api/game/reveal", {method: "POST", body: {x, y}});
    // },
    revealPixels(pixels) {
        return apiFetch("/api/game/reveal", {method: "POST", body: {pixels}});
    },
    submitGuess(guess) {
        return apiFetch("/api/game/submit", {method: "POST", body: {guess}});
    },
    giveUp() {
        return apiFetch("/api/game/giveup", {method: "POST"});
    },
};

File: solution/notebooks/utils/pixcavation_api.py
import requests


class ApiError(Exception):
    pass


class PixcavationAPI:
    def __init__(self, base_url: str):
        self.base_url = base_url.rstrip("/")

    def _fetch(self, path, *, method="GET", body=None):
        url = f"{self.base_url}{path}"

        res = requests.request(
            method,
            url,
            headers={"Content-Type": "application/json"},
            json=body if body is not None else None,
        )

        content_type = res.headers.get("content-type", "")
        data = res.json() if "application/json" in content_type else res.text

        if not res.ok:
            if isinstance(data, dict):
                raise ApiError(data.get("error") or f"Request failed ({res.status_code})")
            raise ApiError(f"Request failed ({res.status_code})")

        if isinstance(data, dict) and data.get("ok") is False:
            raise ApiError(data.get("error") or "Request failed")

        return data

    def exists(self):
        return self._fetch("/api/game/exists")

    def new_game(self):
        return self._fetch("/api/game/new", method="POST")

    def get_game(self):
        return self._fetch("/api/game")

    def reveal_one_pixel(self, x, y):
        return self._fetch(
            "/api/game/reveal",
            method="POST",
            body={"pixels": [{"x": x, "y": y}]},
        )

    def submit_guess(self, guess):
        return self._fetch(
            "/api/game/submit",
            method="POST",
            body={"guess": guess},
        )

    def give_up(self):
        return self._fetch("/api/game/giveup", method="POST")
